nomos-core/chain-defs/src/proofs/leader_proof.rs

```diff
@@ -1,100 +1,279 @@
-            .unwrap();
-
+impl Groth16LeaderProof {
+    pub fn prove(private: &LeaderPrivate, voucher_cm: VoucherCm) -> Result<Self, Error> {
         let start_t = std::time::Instant::now();
+        let (proof, entropy_contribution) = if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being generated in dev mode. This should never be used in production."
+            );
+            (
+                groth16::CompressedGroth16Proof::new(
+                    [0; 32].into(),
+                    [0; 64].into(),
+                    [0; 32].into(),
+                ),
+                Fr::ZERO,
+            )
+        } else {
+            let (proof, verif_inputs) =
+                pol::prove(&private.input).map_err(Error::PoLProofFailed)?;
+            (proof, verif_inputs.entropy_contribution.into_inner())
+        };
+        tracing::debug!("groth16 prover time: {:.2?}", start_t.elapsed(),);
 
-        // ATTENTION: producing a groth16 proof currently requires x86 with docker
-        // support
-        let opts = risc0_zkvm::ProverOpts::groth16();
-        let prove_info =
-            prover.prove_with_opts(env, nomos_risc0_proofs::PROOF_OF_LEADERSHIP_ELF, &opts)?;
+        let leader_key = private.pk;
 
-        tracing::debug!(
-            "STARK prover time: {:.2?}, total_cycles: {}",
-            start_t.elapsed(),
-            prove_info.stats.total_cycles
-        );
-        // extract the receipt.
         Ok(Self {
-            public_inputs,
-            risc0_receipt: prove_info.receipt,
+            proof,
+            entropy_contribution,
+            leader_key,
+            voucher_cm,
         })
     }
+
+    #[must_use]
+    pub const fn proof(&self) -> &pol::PoLProof {
+        &self.proof
+    }
 }
 
 pub trait LeaderProof {
     /// Verify the proof against the public inputs.
     fn verify(&self, public_inputs: &LeaderPublic) -> bool;
 
     /// Get the entropy used in the proof.
-    fn entropy(&self) -> [u8; 32];
+    fn entropy(&self) -> Fr;
+
+    fn leader_key(&self) -> &ed25519_dalek::VerifyingKey;
+
+    fn voucher_cm(&self) -> &VoucherCm;
 }
 
-impl LeaderProof for Risc0LeaderProof {
+impl LeaderProof for Groth16LeaderProof {
     fn verify(&self, public_inputs: &LeaderPublic) -> bool {
-        // The risc0 proof is valid by contract
-        public_inputs == &self.public_inputs
+        if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being verified in dev mode. This should never be used in production."
+            );
+            return true;
+        }
+        let leader_pk_bytes = self.leader_key.as_bytes();
+        let leader_pk = (
+            Fr::from(u128::from_le_bytes(
+                leader_pk_bytes[0..16].try_into().unwrap(),
+            )),
+            Fr::from(u128::from_le_bytes(
+                leader_pk_bytes[16..32].try_into().unwrap(),
+            )),
+        );
```

Comment:
Maybe we could have a type for this? and implement `From<([u8; 16], [u8; 16])>`?